src/java/org/apache/cassandra/cql3/functions/FunctionFactory.java

```diff
@@ -101,9 +112,14 @@ public NativeFunction getOrCreateFunction(List<? extends AssignmentTestable> arg
                 throw new InvalidRequestException(String.format("Cannot infer type of argument %s in call to " +
                                                                 "function %s: use type casts to disambiguate",
                                                                 arg, this));
-            parameter.validateType(name, arg, type);
             type = type.udfType();
-            types.add(type);
+            types.set(i, type);
+        }
+
+        // Validate the inferred types of the arguments, again favouring a left-to-right reading.
+        for (int i = 0; i < args.size(); i++)
+        {
+            parameters.get(i).validateType(name, args.get(i), types.get(i));
```

Comment:
why we validateType them independently here? Can't we move back to line 115? Can you help give me an example about support "left to right", you know we have change the loop about the args from one to three . If there are many parameters(dimention), there may be three times the overhead

Reply:
Indeed we can drop the reverse iteration and have more consistent error messages with some fixes on `FunctionParameter.sameAs`.